File: src/main/java/br/com/blz/testjava/repository/ProductRepositoryImpl.java
package br.com.blz.testjava.repository;

import br.com.blz.testjava.model.ProductEntity;
import org.springframework.stereotype.Repository;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

@Repository
public class ProductRepositoryImpl implements ProductRepository{

    private final Map<Integer,ProductEntity> products = new ConcurrentHashMap<>();

    @Override
    public void createProduct(ProductEntity productEntity) {
        products.put(productEntity.getSku(),productEntity);
    }

    @Override
    public void updateProduct(final ProductEntity productModel) {
        products.put(productModel.getSku(), productModel);
    }

    @Override
    public void deleteProduct(final Integer sku){
        products.remove(sku);
    }

    @Override
    public ProductEntity findProductBySku(final Integer sku){
        return products.get(sku);
    }
}

File: src/main/kotlin/br/com/blz/testjava/repository/ProductRepositoryImpl.kt
package br.com.blz.testjava.repository

import br.com.blz.testjava.cache.ProductCacheContext
import br.com.blz.testjava.model.Product
import org.springframework.stereotype.Repository

@Repository
internal class ProductRepositoryImp : ProductRepository {
  override fun get(sku: Long): Product? {
    return ProductCacheContext.get(sku)
  }

  override fun create(product: Product): Product {
    return ProductCacheContext.save(product)
  }

  override fun update(sku: Long, product: Product): Product {
    val optionalProduct = get(sku)
    optionalProduct?.let { return ProductCacheContext.update(sku, product) } ?: return ProductCacheContext.save(product)
  }

  override fun delete(sku: Long) {
    ProductCacheContext.delete(sku)
  }
}
